### ISSUE TITLE
Command-R GGUF conversion no longer working

### ISSUE THREAD

**Issue by candre23** [ggerganov/llama.cpp#7030]
As recently as a few days ago, Command-R (and presumably R+) could be converted with convert-hf-to-gguf.py.  I double checked and conversion completes successfully in b2751.  However, with the recent changes to accommodate Llama3, Command-R compatibility has been broken.  Trying to convert today with b2777 I get

```
raise NotImplementedError("BPE pre-tokenizer was not recognized - update get_vocab_base_pre()")
NotImplementedError: BPE pre-tokenizer was not recognized - update get_vocab_base_pre()
```

I know that L3 required a new tokenizer provided by meta to facilitate proper conversion.  Do we require something new from cohere, or is this something that can be fixed internally?

**Comment by drummerv** [comment]
I had the same issue and the same error! Had to rollback to before Llama 3 to get it to work.

**Comment by dranger003** [comment]
This patch appears to get it working again.
```
diff --git a/convert-hf-to-gguf.py b/convert-hf-to-gguf.py
index 2f146d73..e8cc040f 100755
--- a/convert-hf-to-gguf.py
+++ b/convert-hf-to-gguf.py
@@ -306,6 +306,8 @@ class Model(ABC):
         if chkhsh == "3ce83efda5659b07b1ad37ca97ca5797ea4285d9b9ab0dc679e4a720c9da7454":
             # ref: https://huggingface.co/openai-community/gpt2
             res = "gpt-2"
+        if chkhsh == "9c2227e4dd922002fb81bde4fc02b0483ca4f12911410dee2255e4987644e3f8":
+            res = "gpt-2"

         if res is None:
             print("\n")
```

**Comment by dranger003** [comment]
There seems to be an issue using `convert-hf-to-gguf-update.py` and the downloaded `tokenizer.json` file does not contain JSON but rather contains this:
```
version https://git-lfs.github.com/spec/v1
oid sha256:0af6e6fe50ce1bb5611b103482de6bac000c82e06898138d57f35af121aec772
size 12777406
```
This is after adding this to `convert-hf-to-gguf-update.py`:
```
{ "name": "command-r",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/CohereForAI/c4ai-command-r-v01", },
```

**Comment by candre23** [comment]
I downloaded the model manually earlier today and the tokenizer.json is definitely a real tokenizer.  The file size mentioned in the placeholder that you got matches the actual size of what I got from cohere's HF repo (~12.1mb).

**Comment by sealad886** [comment]
> There seems to be an issue using `convert-hf-to-gguf-update.py` and the downloaded `tokenizer.json` file does not contain JSON but rather contains this:
> 
> ```
> version https://git-lfs.github.com/spec/v1
> oid sha256:0af6e6fe50ce1bb5611b103482de6bac000c82e06898138d57f35af121aec772
> size 12777406
> ```
> 
> This is after adding this to `convert-hf-to-gguf-update.py`:
> 
> ```
> { "name": "command-r",      "tokt": TOKENIZER_TYPE.BPE, "repo": "https://huggingface.co/CohereForAI/c4ai-command-r-v01", },
> ```

This is caused by Git LFS (for [more details](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage#pointer-file-format)). To fix this, download the `tokenizer.json` file directly from  Huggingface and **replace** the file that was previously downloaded:
<img width="1096" alt="Screenshot 2024-05-02 at 07 37 18" src="https://github.com/ggerganov/llama.cpp/assets/155285242/1d16e2e1-8068-4e1d-a65f-1d4d5b1729a4">
[tokenizer.json for command-r](https://huggingface.co/CohereForAI/c4ai-command-r-v01/resolve/main/tokenizer.json)
--or--
[tokenizer.json for command-r-plus](https://huggingface.co/CohereForAI/c4ai-command-r-plus/resolve/main/tokenizer.json)

**Comment by sealad886** [comment]
> As @candre23 noted, this also affects `command-r-plus`...and every model of a specific type (BPE pre-tokenizer? I'm not super familiar with that). It's a non-trivial but straightforward way to add back model types to `convert-hf-to-gguf.py`. Follow the instructions in [this](https://github.com/ggerganov/llama.cpp/pull/6920#issue-2265280504) pull request. Admittedly, I got a bit lost trying to figure out what regex to use for the `command-r-plus` pre-tokenizer type (it appears to me that Cohere have made this one unified type?). I also might have royallly screwed up my source files accidentally. I was too tired when I started looking at this, and by the end couldn't put one thought in front of another...

Got this working myself about the same time I saw your PR, @dranger003. It does appear this is fixed by #7033 so let's get that merged asap. :)

**Comment by jukofyork** [comment]
Same problem for `dbrx-instruct`:

```
Loading model: dbrx-instruct
gguf: This GGUF file is for Little Endian only
Set model parameters
gguf: file type = 1
Set model tokenizer
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
chktok: [198, 4815, 15073, 66597, 8004, 1602, 2355, 79772, 11187, 9468, 248, 222, 320, 8416, 8, 27623, 114, 378, 235, 9468, 234, 104, 31643, 320, 36773, 100166, 98634, 8, 26602, 227, 11410, 99, 247, 9468, 99, 247, 220, 18, 220, 1644, 220, 8765, 220, 8765, 18, 220, 8765, 1644, 220, 8765, 8765, 220, 8765, 8765, 18, 220, 8765, 8765, 1644, 220, 18, 13, 18, 220, 18, 497, 18, 220, 18, 1131, 18, 220, 21549, 222, 98629, 241, 45358, 233, 21549, 237, 45358, 224, 21549, 244, 21549, 115, 21549, 253, 45358, 223, 21549, 253, 21549, 95, 98629, 227, 76460, 223, 949, 37046, 33565, 111, 19000, 23182, 49792, 19967, 9263, 18136, 16, 36827, 21909, 56560, 54337, 19175, 14476, 1482, 13373, 64571, 34694, 3114, 15752, 17721, 80112, 3436, 4708, 4708, 14196, 14196, 74694, 3089, 3089, 29249, 17523, 3001, 27708, 7801, 358, 3077, 1027, 364, 83, 820, 568, 596, 1070, 11, 364, 793, 499, 2771, 30, 364, 44, 539, 2771, 358, 3358, 1304, 433, 11, 364, 35, 499, 1093, 1063, 15600, 30, 1226, 6, 43712, 264, 64966, 43]
chkhsh: a8594e3edff7c29c003940395316294b2c623e09894deebbc65f33f1515df79e


**************************************************************************************
** WARNING: The BPE pre-tokenizer was not recognized!
**          There are 2 possible reasons for this:
**          - the model has not been added to convert-hf-to-gguf-update.py yet
**          - the pre-tokenization config has changed upstream
**          Check your model files and convert-hf-to-gguf-update.py and update them accordingly.
** ref:     https://github.com/ggerganov/llama.cpp/pull/6920
**
** chkhsh:  a8594e3edff7c29c003940395316294b2c623e09894deebbc65f33f1515df79e
**************************************************************************************


Traceback (most recent call last):
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 3001, in <module>
    main()
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 2988, in main
    model_instance.set_vocab()
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 75, in set_vocab
    self._set_vocab_gpt2()
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 331, in _set_vocab_gpt2
    tokens, toktypes, tokpre = self.get_vocab_base()
                               ^^^^^^^^^^^^^^^^^^^^^
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 242, in get_vocab_base
    tokpre = self.get_vocab_base_pre(tokenizer)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/juk/LLMs/llama.cpp/convert-hf-to-gguf.py", line 323, in get_vocab_base_pre
    raise NotImplementedError("BPE pre-tokenizer was not recognized - update get_vocab_base_pre()")
NotImplementedError: BPE pre-tokenizer was not recognized - update get_vocab_base_pre()
main: build = 2776 (c4ec9c0d)
main: built with cc (Debian 12.2.0-14) 12.2.0 for x86_64-linux-gnu
```
